06-TimeManagement/omod6_kern_time.c

```diff
@@ -0,0 +1,204 @@
+// - return absolute time of previous reading;
+// - periodically print some message to log, make period adjustable;
+
+// --------------- MODULE PARAMETERS DESCRIPTION -------------------
+//int iParam;
+//module_param(iParam, int, 0);
+//MODULE_PARM_DESC(iParam, "iParam: ReturnCode");
+
+//char *sParam = "Default";
+//module_param(sParam, charp, 0);
+//MODULE_PARM_DESC(sParam, "sParam: string parameter");
+
+//------------ GLOBAL MODULE DATA ---------------------
+static struct class *attr_class;
+
+#define SysFS_DIR_NAME   "omod6"
+#define SysFS_ATTR_NAME  "ktime"
+
+uint32_t ModCallCount;
+uint32_t LastCallTimeSec;
+uint32_t LastCall_dT;
+uint32_t LastCallTicks;
+
+uint32_t TimerInterval;
+uint32_t TimerCounter;
+uint32_t TimerStartTick;
+
+char LogMessage[PAGE_SIZE];
+
+uint32_t GetTimeStamp(void)
+{
+	return (uint32_t) (ktime_to_ms(ktime_get_real()) / 1000);
+}
+
+uint32_t GetTickCount(void)
+{
+	return (uint32_t) (get_jiffies_64() * 1000 / HZ);
+}
+
+
+//--------------- SysFS read/write functions ----------------------
+
+static ssize_t ktime_show(struct class *cl,
+			struct class_attribute *attr,
+			char *buf)
+{
+	int L;
+
+	uint32_t TS, dT, T;
+
+	T = GetTickCount();
+	TS = GetTimeStamp();
+	dT = TS - LastCallTimeSec;
+
+	if (dT == 0)
+		dT = LastCall_dT;
+	else
+		LastCall_dT = dT;
+
+	LastCallTicks = T;
+
+	ModCallCount++;
+
+	sprintf(buf, "ktime (%u ticks, %u HZ):\n", (uint32_t)get_jiffies_64(), HZ);
+	sprintf(strchr(buf, 0), "%u module call\n", ModCallCount);
+	sprintf(strchr(buf, 0), "%u sec from last call\n", dT);
+	sprintf(strchr(buf, 0), "%u sec curr abs time from Epoch (year:%d)\n", TS, 1970+TS/(3600*24*365));
+	sprintf(strchr(buf, 0), "%u sec prev call abs time from Epoch\n", LastCallTimeSec);
+	sprintf(strchr(buf, 0), "%u ms - Timer Interval\n", TimerInterval);
+	sprintf(strchr(buf, 0), "%u times timer called\n", TimerCounter);
+
+	LastCallTimeSec = TS;
+
+	L = strlen(buf);
+	return L;
+}
+
+// Kernel Timer object
+struct timer_list MyTimer;
+
+//
+// Kernel Timer Function. Run in interrupt context! Called only once after start/mod
+//
+void KernelTimerFunc(struct timer_list *t)
+{
+	uint32_t T, dT;
+
+	TimerCounter++;
+	T = GetTickCount();
+	dT = T - TimerStartTick;
+	pr_info("omod6 timer[%u]: %s (delay = %u ms)\n",
+		TimerCounter, LogMessage, dT);
+}
+
+
+static ssize_t ktime_store(struct class *cl,
+			struct class_attribute *attr,
+			const char *buf, size_t count)
+{
+	int interval = 0;
+	int n;
+
+	n = sscanf(buf, "%d %s", &interval, LogMessage);
+	pr_info("omod6 store (%d, \'%s\')\n", interval, LogMessage);
+	if (n >= 1) {
+		TimerInterval = interval;
+		// start, stop or reconfig timer
+		if (interval == 0) {
+			if (timer_pending(&MyTimer)) {
+				del_timer(&MyTimer);
+				pr_info("omod6 Timer disabled\n");
+			}
+		} else {
+			unsigned long expired = get_jiffies_64() +
+				interval * HZ / 1000;
+			//if (timer_pending(&MyTimer)) {
+			//	mod_timer_pending(&MyTimer, expired);
+			//} else {
```

Comment:
What is the purpose of these lines commented out?

Reply:
1. I had to do some experiments and kernel code inspection to clarify the behavior of timers and timer functions. A lot of lecture information about timers was unclear, obsolete and non-actual.
2. That commented code has been left to accentuate that used functions was tested and recognized as excessive. Everything works perfect without them. But it works with them, too.